### PR TITLE
JSONP doesn't support async: false, so fails with redirecting

### PR DESCRIPTION
**Pull request by iHiD**
There are currently sync problems with JSONP as the router redirect fires before the request comes back. The most robust way I've found to fix this is to change the redirect section to:

``` javascript
  Auth.Route.reopen({
    redirect: function() {
      if (Auth.Config.get('rememberMe') && Auth.Config.get('rememberAutoRecall')) {
        request = Auth.Module.RememberMe.recall({async: true}) 
        if(request) {
          window.Meducation.theRequest = request
          self = this;
          callback = this._super;
          return request.always(function() {
            callback.call(self)
          });
        }
      }
      return this._super();
    }
  });
```

I've also changed to the new deferred jQuery methods, as the old callbacks don't play nicely with JSONP either (e.g. no error callback).

I've tested this lots with JSONP and it's working great now, but you might like to give it a run through on standard JSON, just in case I've missed anything. 

This contains the same functionality as my other pull request, so if you pull this, you can ignore the other, but if you want to discuss this in more detail, the other one is a nice quick fix. Thanks! :smiley_cat:

### CODE REVIEW

**Comment by iHiD** [comment]
I'm also happy to update the callbacks for DELETE to promises too, if you like, to keep things consistent. Let me know!

**Comment by iHiD** [comment]
Just realised that I've edited the lib, not the src. Fail! :disappointed:

If you like the ideas, I'll edit the source for you. If not, I won't bother! :) Let me know.

**Comment by heartsentwined** [comment]
I welcome this idea very much, but what does this line do?

``` coffeescript
window.Meducation.theRequest = request
```

Google yields nothing, so I guess that's your app? (Not acceptable in this case...) I am keeping #15 open if this guess is true, or unless we can find another robust method here.

**Comment by iHiD** [comment]
Oh, that was just there for testing. Sorry. It just exposed the request onto my App's namespace (Meducation) so I could play around with it. It can be happily removed.

OK, I'll port this (and #15) to coffeescript today or tomorrow and send over an updated PR for you! Thanks :)

**Comment by iHiD** [comment]
I've updated this too. It should merge independently of the others. I've only checked it with JSONP, not JSON, so please give it a good once over to make sure I've not broken anything :)

Thanks again!

File: src/modules/remember-me.coffee [hunk 1]
```diff
@@ -39,5 +39,9 @@ Auth.Module.RememberMe = Em.Object.create
 Auth.Route.reopen
   redirect: ->
     if Auth.Config.get('rememberMe') && Auth.Config.get('rememberAutoRecall')
-      Auth.Module.RememberMe.recall { async: false }
+      if request = Auth.Module.RememberMe.recall(async: false)
+        self = @
+        callback = @_super
+        return request.always ->
+          callback.call(self)
```
Comment (heartsentwined) [inline src/modules/remember-me.coffee:46]: I believe that coffeescript will take care of this `this` hack automatically with the fat arrow operator, such that this section can be reduced to

``` coffeescript
if request = Auth.Module.RememberMe.recall(async: false)
  return request.always => @_super()
```

Give it a try.

I could've just fixed this on merging but, this is OSS and this is, of course, your credit. :-)

Comment (iHiD) [inline src/modules/remember-me.coffee:46]: Ha. Thanks :)

I actually think you need the little dance I've got there. When I change the code what you have, I get `Uncaught TypeError: Property '_super' of object [object Object] is not a function` 

I also tried `return request.always => @_super.call(@)` but that fails too with `Uncaught TypeError: Cannot call method 'call' of undefined`.

I believe both the receiver and the context need to be set in order to call `_super()` - which requires storing the method and then using `call`. Saying that, I'm extremely happy for you to show me a better way if you can find one :)

Comment (heartsentwined) [inline src/modules/remember-me.coffee:46]: Alright, I'll merge the current code now and refactor it later.

Comment (iHiD) [inline src/modules/remember-me.coffee:46]: Great - thanks. Can you let me know when you refactor it. I'd like to see how you get round it. Thanks :)

**Comment by heartsentwined** [comment]
Sure. I'm building a test suite for `ember-auth` at the moment, so it'll be after that.

**Comment by heartsentwined** [comment]
Test suite released in `3.0` branch, but I have not included `jsonp` tests yet because I have not used it before.

If you have time, would you mind writing the corresponding `jsonp` tests for this PR? Or perhaps give me some pointers as to how `jsonp` requests are formatted? The only thing I know is that they are usually wrapped in a function, but perhaps there are other `content-type`, `header`, etc, issues?

I have used [`jasmine`](http://pivotal.github.com/jasmine/) and [`$.mockjax`](https://github.com/appendto/jquery-mockjax) for the tests.

**Pull request merged**